test/libyul/yulInterpreterTests/and_create.yul

```diff
@@ -0,0 +1,13 @@
+{
+    let u160max := 0xffffffffffffffffffffffffffffffffffffffff
+    let a := create(0, u160max, 0)
```

Comment:
Not sure it is a good idea to use `u160max` as a memory offset.

Reply:
I'm afraid we need to use a large number as memory offset because of the following:

- When the Yul interpreter encounters a `create` (likewise `create2`) instruction, the following lines are executed

https://github.com/ethereum/solidity/blob/e970e58a9fe813e1c791f295be97553036c06f7a/test/tools/yulInterpreter/EVMInstructionInterpreter.cpp#L303-L310

- The interpreter returns an address whose value is `0xcccccc` (respectively `0xdddddd` for `create2`) plus memory offset

- However, the "bug" in the interpreter is that the returned value is of type u256 which means that memory offsets that would otherwise overflow u160 type (e.g., `u160max - 0xcccccc >= offset`) will simply be interpreted as large u256 values i.e., the returned "address" is actually a value `> u160max`.

- This meant that, earlier (pre fix) when memory offset is large (e.g., `u160max`) the interpreter would return different values pre and post optimisation because of the rule that `and(u160max, create(...)) == create(...)`

- Now that this bug has been fixed by casting the return value to a `u160` type, this test ensures that `u160` overflows are interpreted correctly.

Reply:
What `create` should do is hash the contents of the memory and use that as the address. That is roughly what create2 does, but I think would be okay for both create and create2.

Then you don't have this problem and it is more realistic.

Reply:
> What `create` should do is hash the contents of the memory and use that as the address.

Sounds like a separate issue to me :) I have a feeling hashing would slightly slow down the interpreter and hence fuzzing but that needs to be seen.

Reply:
Could even just use adding up the bytes, no need for real cryptographic hashes.

However, `accessMemory` should either try to expand memory or calculate gas and both cases would fail if you try to expand it to `u160max`.

Reply:
> However, `accessMemory` should either try to expand memory

IIUC, `accessMemory` does "expand" memory even for large create memory offset (say `u160max`). See

https://github.com/ethereum/solidity/blob/8226f3e80464e83c7763ee6ae7dc434c417b9181/test/tools/yulInterpreter/EVMInstructionInterpreter.cpp#L463

Please note that `msize` is of type `u256` so passing `u160max` as memory offset works just fine.

>  or calculate gas and both cases would fail if you try to expand it to `u160max`.

Afaik, gas calculation is not modeled in the yul interpreter yet. Please see https://github.com/ethereum/solidity/issues/7585#issuecomment-547826759